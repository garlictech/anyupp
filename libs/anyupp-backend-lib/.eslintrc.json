{
  "extends": "../../.eslintrc.json",
  "ignorePatterns": [
    "!**/*"
  ],
  "rules": {},
  "overrides": [
    {
      "files": [
        "*.ts"
      ],
      "plugins": [
        "@typescript-eslint"
      ],
      "parserOptions": {
        "project": [
          "libs/anyupp-backend-lib/tsconfig.*?.json"
        ]
      }
    }
  ]
}
