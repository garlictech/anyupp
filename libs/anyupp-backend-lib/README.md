# anyupp-backend-lib

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test anyupp-backend-lib` to execute the unit tests via [Jest](https://jestjs.io).
