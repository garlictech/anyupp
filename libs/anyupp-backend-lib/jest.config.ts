module.exports = {
  displayName: 'anyupp-backend-lib',

  globals: {
    'ts-jest': { tsconfig: '<rootDir>/tsconfig.spec.json' },
  },
  transform: {
    '^.+\\.[tj]sx?$': 'ts-jest',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../coverage/libs/anyupp-backend-lib',
  preset: '../../jest.preset.ts',
};
