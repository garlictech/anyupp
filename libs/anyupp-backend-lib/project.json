{
  "root": "libs/anyupp-backend-lib",
  "sourceRoot": "libs/anyupp-backend-lib/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "options": {
        "lintFilePatterns": ["libs/anyupp-backend-lib/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/libs/anyupp-backend-lib"],
      "options": {
        "jestConfig": "libs/anyupp-backend-lib/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": ["platform:nodejs", "type:backend"]
}
