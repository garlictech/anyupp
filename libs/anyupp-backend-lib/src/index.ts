export * from './lib/report';
export * from './lib/seeder';
export * from './lib/infra';
export * from './lib/database';
export * from './lib/date/date.utils';
export * from './lib/resolver-utils/get-authenticated-userid';
