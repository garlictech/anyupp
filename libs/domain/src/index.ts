export * from './lib/use-cases';
export * from './lib/entities';
