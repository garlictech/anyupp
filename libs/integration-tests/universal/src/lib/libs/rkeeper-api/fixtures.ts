import * as CrudApi from '@bgap/crud-gql/api';
import {
  Dish,
  RKeeperBusinessEntityInfo,
  RkeeperFixtures,
} from '@bgap/rkeeper-api';
import { RequiredId } from '@bgap/shared/types';
import {
  unitFixture,
  productFixture,
  chainFixture,
  groupFixture,
  rkeeperEndpoint,
  yellowRestaurantId,
  yellowRkeeperUsername,
  yellowRkeeperPassword,
} from '@bgap/shared/fixtures';

const testIdPrefix = `rkeeper-cf0d1110-a2ce-45cf-aa69-6782bbc44cad`;

export const rawData = {
  data: {
    dishes: [
      RkeeperFixtures.dish,
      RkeeperFixtures.dish2,
      RkeeperFixtures.duplicatedDish,
      RkeeperFixtures.badDish,
      RkeeperFixtures.inactiveDish,
      RkeeperFixtures.dishWithModifier,
      RkeeperFixtures.dishWithSameModifier,
    ],
    modifiers: [RkeeperFixtures.modifier, RkeeperFixtures.unusedModifier],
  },
};

export const rkeeperProductGuid = 'RKEEPERGUID';

const unitId = `${testIdPrefix}-unit`;
const chainId = `${testIdPrefix}-chain`;
const groupId = `${testIdPrefix}-group`;

export const rkeeperUnit: RequiredId<CrudApi.CreateUnitInput> = {
  ...unitFixture.createRkeeperUnit,
  id: unitId,
  groupId,
  chainId,
  externalId: 'EXTERNAL-RESTAURANT-ID',
};

export const rkeeperUnitProduct: RequiredId<CrudApi.CreateUnitProductInput> = {
  ...productFixture.unitProductInputBase,
  id: `${testIdPrefix}-unitproduct`,
  externalId: rkeeperProductGuid,
};

export const rkeeperUnitProduct2: RequiredId<CrudApi.CreateUnitProductInput> = {
  ...productFixture.unitProductInputBase,
  id: `${testIdPrefix}-unitproduct2`,
  externalId: rkeeperProductGuid + '2',
};

export const processedDish: Dish = {
  price: 50000,
  active: true,
  id: 9991000114,
  guid: '999e3ab3-86a0-48d9-a9a9-f4e0c9fbce68',
  name: 'Pr√≥ba ital',
};

export const businessEntity: RKeeperBusinessEntityInfo = {
  chainId,
  groupId,
  unitId,
};

export const createChain: RequiredId<CrudApi.CreateChainInput> = {
  ...chainFixture.chainBase,
  id: chainId,
};

export const createGroup: RequiredId<CrudApi.CreateGroupInput> = {
  ...groupFixture.groupBase,
  id: groupId,
  chainId,
};

export const rkeeperOrder = {
  objectid: yellowRestaurantId,
  remoteId: yellowRestaurantId,
  order_type: 1,
  pay_type: 0,
  pay_online_type: 0,
  delivery_time: '2122-02-03T12:37:32',
  client: {
    phone: null,
    email: null,
    ln: 'Testln',
    fn: 'Testfn',
  },
  order: [
    {
      id: '1000114',
      type: 'd',
      qnt: 1000,
    },
  ],
};

export const yellowUnit = {
  ...rkeeperUnit,
  externalId: yellowRestaurantId,
  pos: {
    ...rkeeperUnit.pos,
    rkeeper: {
      ...(rkeeperUnit?.pos?.rkeeper || {}),
      rkeeperUsername: yellowRkeeperUsername,
      rkeeperPassword: yellowRkeeperPassword,
      endpointUri: rkeeperEndpoint,
    } as CrudApi.RKeeper,
  },
} as CrudApi.Unit;

export const orderInput: CrudApi.CreateOrderInput = {
  userId: 'USER_ID',
  unitId: yellowUnit.id,
  items: [
    {
      quantity: 5,
      productId: 'PRODUCT ID',
      statusLog: [
        {
          userId: 'USER_ID',
          status: CrudApi.OrderStatus.none,
          ts: 1627909024677,
        },
      ],
      sumPriceShown: {
        taxSum: 316.98,
        currency: 'HUF',
        tax: 27,
        priceSum: 1491,
        pricePerUnit: 298.2,
      },
      productName: {
        hu: 'PRODUCT NAME HU',
      },
      priceShown: {
        taxSum: 318.9,
        currency: 'HUF',
        tax: 27,
        priceSum: 1500,
        pricePerUnit: 300,
      },
      variantId: 'VARIANT ID',
      variantName: {
        en: 'VARIANT NAME EN',
      },
    },
  ],
  sumPriceShown: {
    taxSum: 633.96,
    currency: 'HUF',
    tax: 27,
    priceSum: 2982,
    pricePerUnit: 298.2,
  },
  place: {
    table: '01',
    seat: '01',
  },
  orderMode: CrudApi.OrderMode.instant,
  servingMode: CrudApi.ServingMode.inplace,
  archived: false,
  paymentMode: {
    method: CrudApi.PaymentMethod.cash,
    type: CrudApi.PaymentType.cash,
  },
};

export {
  yellowRestaurantId,
  yellowRkeeperPassword,
  yellowRkeeperUsername,
  rkeeperEndpoint,
};
