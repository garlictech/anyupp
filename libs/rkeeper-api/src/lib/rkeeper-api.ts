import { AWSError, ECS } from 'aws-sdk';
import { pipe } from 'fp-ts/lib/function';
import { bindNodeCallback, defer } from 'rxjs';
import { map, switchMap, tap } from 'rxjs/operators';
import { S3 } from 'aws-sdk';

export interface HandleProductsDeps {
  ecs: ECS;
  RKeeperProcessProductSubnet: string;
  RKeeperProcessProductSecurityGroup: string;
  taskDefinitionArn: string;
  bucketName: string;
  uuidGenerator: () => string;
}

export const handleProducts =
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (deps: HandleProductsDeps) => (unitId: string, rawData: any) => {
    const objectKey = deps.uuidGenerator();
    const s3 = new S3();

    return pipe(
      JSON.stringify(rawData),
      Body => ({
        Bucket: deps.bucketName,
        Key: objectKey,
        Body,
      }),
      params => defer(() => s3.upload(params).promise()),
      tap(x => console.log('*****1', x)),
      map(() => ({
        launchType: 'FARGATE',
        networkConfiguration: {
          awsvpcConfiguration: {
            subnets: [deps.RKeeperProcessProductSubnet],
            securityGroups: [deps.RKeeperProcessProductSecurityGroup],
          },
        },
        taskDefinition: deps.taskDefinitionArn,
        cluster: 'anyupp-fargate-cluster',
        overrides: {
          containerOverrides: [
            {
              name: 'DefaultContainer',
              environment: [
                {
                  name: 'unitId',
                  value: unitId,
                },
                {
                  name: 'objectKey',
                  value: objectKey,
                },
              ],
            },
          ],
        },
      })),
      tap(x => console.log('*****2', x)),
      switchMap(params =>
        bindNodeCallback(
          (
            p: ECS.Types.RunTaskRequest,
            callback: (err: AWSError, data: ECS.Types.RunTaskResponse) => void,
          ) => deps.ecs.runTask(p, callback),
        )(params),
      ),
      tap(x => console.log('*****3', x)),
      tap(result =>
        console.log(
          'Task submission result: ',
          JSON.stringify(result, null, 2),
        ),
      ),
    );
  };
