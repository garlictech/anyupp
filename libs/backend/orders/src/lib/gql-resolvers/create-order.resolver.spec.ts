/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable @typescript-eslint/no-explicit-any */

import { of, throwError } from 'rxjs';
import { catchError, tap } from 'rxjs/operators';
import {
  CalculationState_UnitAdded,
  getUnit,
  getNextOrderNum,
  placeOrder,
  handleRkeeperOrder,
  CalculationState_OrderAdded,
  createOrder,
} from './create-order.resolver';
import * as anyuppBackendLib from '@bgap/anyupp-backend-lib';
import * as rkeeperApi from '@bgap/rkeeper-api';
import { CreateOrderInput, OrderPaymentPolicy, PosType } from '@bgap/domain';

jest.mock('@bgap/rkeeper-api', () => ({
  __esModule: true,
  // @ts-ignore
  ...jest.requireActual('@bgap/rkeeper-api'),
}));

jest.mock('@bgap/anyupp-backend-lib', () => ({
  __esModule: true,
  // @ts-ignore
  ...jest.requireActual('@bgap/anyupp-backend-lib'),
}));

Date.now = () => 1627909024677;

const getUnitCases = [
  { orderInput: { unitId: 'UNIT ID' }, unit: of(undefined) },
  { orderInput: { unitId: 'UNIT ID' }, unit: throwError('UNIT ERROR') },
  { orderInput: { unitId: 'UNIT ID' }, unit: of({ id: 'UNIT ID 1' }) },
  {
    orderInput: { unitId: 'UNIT ID' },
    unit: of({ id: 'UNIT ID 2', isAcceptingOrders: true }),
  },
  {
    orderInput: { unitId: 'UNIT ID' },
    unit: of({
      id: 'UNIT ID 3',
      isAcceptingOrders: true,
      orderPaymentPolicy: OrderPaymentPolicy.afterpay,
    }),
  },
  {
    orderInput: { unitId: 'UNIT ID' },
    unit: of({
      id: 'UNIT ID 4',
      isAcceptingOrders: true,
      orderPaymentPolicy: OrderPaymentPolicy.prepay,
    }),
  },
  {
    orderInput: { unitId: 'UNIT ID', paymentMode: {} },
    unit: of({
      id: 'UNIT ID 5',
      isAcceptingOrders: true,
    }),
  },
];

test.each(getUnitCases)('getUnit cases', ({ orderInput, unit }, done: any) => {
  const depsFixture = {
    crudSdk: {
      GetUnit: jest.fn().mockReturnValue(unit),
    },
  } as any;

  getUnit(depsFixture)(orderInput as CreateOrderInput)
    .pipe(
      tap(res => {
        expect(depsFixture.crudSdk.GetUnit.mock.calls).toMatchSnapshot(
          'GETUNIT CALLS',
        );
        expect(res).toMatchSnapshot('RESULT');
      }),
    )
    .subscribe(() => done());
});

const getNextOrderNumCases = [
  {
    inputState: {
      order: { place: { table: 'TABLE1', seat: 'SEAT1' } },
      unit: { id: 'UNIT ID 1' },
    },
    result: of(undefined),
  },
  {
    inputState: {
      order: { place: { table: 'TABLE2', seat: 'SEAT2' } },
      unit: { id: 'UNIT ID 2' },
    },
    result: throwError('INCREMENT ORDER NUM FAILURE'),
  },
  {
    inputState: {
      order: { place: { table: 'TABLE3', seat: 'SEAT3' } },
      unit: { id: 'UNIT ID 3' },
    },
    result: of(10),
  },
  {
    inputState: {
      order: {},
      unit: { id: 'UNIT ID 4' },
    },
    result: of(20),
  },
];

test.each(getNextOrderNumCases)(
  'getNextOrderNum cases',
  ({ inputState, result }, done: any) => {
    const depsFixture = {
      crudSdk: {
        GetGroupCurrency: jest.fn().mockReturnValue(result),
      },
      random: () => 0.5,
      unitTableName: 'UNIT TABLE NAME',
    } as any;

    const funcSpy = jest.fn().mockReturnValue(result);
    const incrementOrderNumSpy = jest
      .spyOn(anyuppBackendLib, 'incrementOrderNum')
      .mockReturnValue(funcSpy);

    getNextOrderNum(depsFixture)(inputState as CalculationState_UnitAdded)
      .pipe(
        tap(res => {
          expect(incrementOrderNumSpy.mock.calls).toMatchSnapshot(
            'INCREMENTORDERNUM DEPS CALLS',
          );
          expect(funcSpy.mock.calls).toMatchSnapshot(
            'INCREMENTORDERNUM FUNC CALLS',
          );
          expect(res).toMatchSnapshot('RESULT');
        }),
      )
      .subscribe(() => done());
  },
);

const placeOrderCases = [
  {
    inputState: {
      order: { unitId: 'THE ORDER UNIT ID' },
    },
    result: Promise.resolve({
      Attributes: { unitId: 'THE ORDER UNIT ID IN ATTRIBUTES' },
    }),
    label: 'All good',
  },
  {
    inputState: {
      order: { unitId: 'THE ORDER UNIT ID' },
    },
    result: Promise.reject('DB ERRORED'),
    label: 'Error happened',
  },
  {
    inputState: {
      order: { unitId: 'THE ORDER UNIT ID', id: 'USER GIVEN ORDER ID' },
    },
    result: Promise.resolve({
      Attributes: { unitId: 'THE ORDER UNIT ID IN ATTRIBUTES' },
    }),
    label: 'Order ID specified',
  },
];

test.each(placeOrderCases)(
  'placeOrderCases cases',
  ({ inputState, result, label }, done: any) => {
    const depsFixture = {
      docClient: {
        put: jest.fn().mockReturnValue({ promise: () => result }),
      },
      uuid: () => 'THE UUID',
      orderTableName: 'ORDER TABLE NAME',
      currentTimeISOString: () => new Date('2022-02-01').toISOString(),
      uuidGenerator: () => 'UUID',
    } as any;

    placeOrder(depsFixture)(inputState as CalculationState_UnitAdded)
      .pipe(
        tap(res => {
          expect(depsFixture.docClient.put.mock.calls).toMatchSnapshot(
            `Case: "${label}", PUT CALLS`,
          );

          expect(res).toMatchSnapshot(`Case: "${label}", RESULT`);
        }),
      )
      .subscribe(() => done());
  },
);

const handleRkeeperOrderCases = [
  {
    inputState: {
      unit: { id: 'NOT AN RKEEPER UNIT' },
      order: { unitId: 'THE ORDER UNIT ID' },
    },
    result: of({}),
  },
  {
    inputState: {
      unit: {
        id: 'AN RKEEPER UNIT',
        pos: {
          type: PosType.rkeeper,
        },
      },
      order: { unitId: 'THE ORDER UNIT ID' },
    },
    result: of({}),
  },
  {
    inputState: {
      unit: {
        id: 'AN RKEEPER UNIT',
        pos: {
          type: PosType.rkeeper,
        },
      },
      order: { unitId: 'THE ORDER UNIT ID' },
    },
    result: throwError('RKEEPER ERROR'),
  },
];

xtest.each(handleRkeeperOrderCases)(
  'handleRkeeperOrderCases cases',
  ({ inputState, result }, done: any) => {
    const depsFixture = {
      currentTimeISOString: () => new Date('2022-02-01').toISOString(),
      axiosInstance: {},
      uuid: () => 'THE UUID',
    } as any;

    const funcSpy = jest.fn().mockReturnValue(result);
    const rkeeperSpy = jest
      .spyOn(rkeeperApi, 'sendRkeeperOrder')
      .mockReturnValue(funcSpy);

    handleRkeeperOrder(depsFixture)(inputState as CalculationState_OrderAdded)
      .pipe(
        tap(res => {
          expect(rkeeperSpy.mock.calls).toMatchSnapshot('RKEEPER DEPS CALLS');
          expect(funcSpy.mock.calls).toMatchSnapshot('RKEEPER FUNC CALLS');
          expect(res).toMatchSnapshot('RESULT');
        }),
      )
      .subscribe(() => done());
  },
);

const createOrderCases = [
  { orderInput: { unitId: 'UNIT ID' }, result: throwError('INCREMENT ERROR') },
  { orderInput: { unitId: 'UNIT ID' }, result: of(10) },
];

test.each(createOrderCases)(
  'createOrderCases cases',
  ({ orderInput, result }, done: any) => {
    const depsFixture = {
      crudSdk: {
        GetUnit: jest.fn().mockReturnValue(
          of({
            id: 'THE UNTI ID',
            isAcceptingOrders: true,
            orderPaymentPolicy: OrderPaymentPolicy.afterpay,
          }),
        ),
      },
      random: () => 0.5,
      unitTableName: 'UNIT TABLE NAME',
      currentTimeISOString: () => new Date('2022-02-01').toISOString(),
      axiosInstance: {},
      uuid: () => 'THE UUID',
      docClient: {
        put: jest.fn().mockReturnValue({
          promise: () =>
            Promise.resolve({
              Attributes: { unitId: 'THE ORDER UNIT ID IN ATTRIBUTES' },
            }),
        }),
      },
    } as any;

    const incrementFuncSpy = jest.fn().mockReturnValue(result);
    jest
      .spyOn(anyuppBackendLib, 'incrementOrderNum')
      .mockReturnValue(incrementFuncSpy);

    const rkeeperFuncSpy = jest.fn().mockReturnValue(of({}));
    jest.spyOn(rkeeperApi, 'sendRkeeperOrder').mockReturnValue(rkeeperFuncSpy);

    createOrder(orderInput as CreateOrderInput)(depsFixture)
      .pipe(
        tap(res => {
          expect(res).toMatchSnapshot('RESULT');
        }),
        catchError(err => {
          expect(err).toMatchSnapshot('ERROR RESULT');
          return of({});
        }),
      )
      .subscribe(() => done());
  },
);
