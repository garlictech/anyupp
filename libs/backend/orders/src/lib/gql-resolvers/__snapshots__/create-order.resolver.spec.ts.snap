// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createOrderCases cases: ERROR RESULT 1`] = `"INCREMENT ERROR"`;

exports[`createOrderCases cases: RESULT 1`] = `
Object {
  "archived": false,
  "createdAt": "2022-02-01T00:00:00.000Z",
  "currentStatus": "none",
  "id": "THE UUID",
  "orderNum": "010",
  "statusLog": Array [
    Object {
      "status": "none",
      "ts": 1627909024677,
      "userId": undefined,
    },
  ],
  "unitId": "UNIT ID",
  "updatedAt": "2022-02-01T00:00:00.000Z",
}
`;

exports[`getNextOrderNum cases: INCREMENTORDERNUM DEPS CALLS 1`] = `
Array [
  Array [
    "UNIT TABLE NAME",
  ],
]
`;

exports[`getNextOrderNum cases: INCREMENTORDERNUM DEPS CALLS 2`] = `
Array [
  Array [
    "UNIT TABLE NAME",
  ],
]
`;

exports[`getNextOrderNum cases: INCREMENTORDERNUM DEPS CALLS 3`] = `
Array [
  Array [
    "UNIT TABLE NAME",
  ],
]
`;

exports[`getNextOrderNum cases: INCREMENTORDERNUM DEPS CALLS 4`] = `
Array [
  Array [
    "UNIT TABLE NAME",
  ],
]
`;

exports[`getNextOrderNum cases: INCREMENTORDERNUM FUNC CALLS 1`] = `
Array [
  Array [
    "UNIT ID 1",
  ],
]
`;

exports[`getNextOrderNum cases: INCREMENTORDERNUM FUNC CALLS 2`] = `
Array [
  Array [
    "UNIT ID 2",
  ],
]
`;

exports[`getNextOrderNum cases: INCREMENTORDERNUM FUNC CALLS 3`] = `
Array [
  Array [
    "UNIT ID 3",
  ],
]
`;

exports[`getNextOrderNum cases: INCREMENTORDERNUM FUNC CALLS 4`] = `
Array [
  Array [
    "UNIT ID 4",
  ],
]
`;

exports[`getNextOrderNum cases: RESULT 1`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "order": Object {
      "orderNum": "050",
      "place": Object {
        "seat": "SEAT1",
        "table": "TABLE1",
      },
    },
    "unit": Object {
      "id": "UNIT ID 1",
    },
  },
}
`;

exports[`getNextOrderNum cases: RESULT 2`] = `
Object {
  "_tag": "Left",
  "left": "INCREMENT ORDER NUM FAILURE",
}
`;

exports[`getNextOrderNum cases: RESULT 3`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "order": Object {
      "orderNum": "010",
      "place": Object {
        "seat": "SEAT3",
        "table": "TABLE3",
      },
    },
    "unit": Object {
      "id": "UNIT ID 3",
    },
  },
}
`;

exports[`getNextOrderNum cases: RESULT 4`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "order": Object {
      "orderNum": "020",
    },
    "unit": Object {
      "id": "UNIT ID 4",
    },
  },
}
`;

exports[`getUnit cases: GETUNIT CALLS 1`] = `
Array [
  Array [
    Object {
      "id": "UNIT ID",
    },
  ],
]
`;

exports[`getUnit cases: GETUNIT CALLS 2`] = `
Array [
  Array [
    Object {
      "id": "UNIT ID",
    },
  ],
]
`;

exports[`getUnit cases: GETUNIT CALLS 3`] = `
Array [
  Array [
    Object {
      "id": "UNIT ID",
    },
  ],
]
`;

exports[`getUnit cases: GETUNIT CALLS 4`] = `
Array [
  Array [
    Object {
      "id": "UNIT ID",
    },
  ],
]
`;

exports[`getUnit cases: GETUNIT CALLS 5`] = `
Array [
  Array [
    Object {
      "id": "UNIT ID",
    },
  ],
]
`;

exports[`getUnit cases: GETUNIT CALLS 6`] = `
Array [
  Array [
    Object {
      "id": "UNIT ID",
    },
  ],
]
`;

exports[`getUnit cases: GETUNIT CALLS 7`] = `
Array [
  Array [
    Object {
      "id": "UNIT ID",
    },
  ],
]
`;

exports[`getUnit cases: RESULT 1`] = `
Object {
  "_tag": "Left",
  "left": "Unit UNIT ID in the cart cannot be fetched from the database.",
}
`;

exports[`getUnit cases: RESULT 2`] = `
Object {
  "_tag": "Left",
  "left": "UNIT ERROR",
}
`;

exports[`getUnit cases: RESULT 3`] = `
Object {
  "_tag": "Left",
  "left": "Unit does not accept orders",
}
`;

exports[`getUnit cases: RESULT 4`] = `
Object {
  "_tag": "Left",
  "left": "Payment mode is not provided in the order, and the unit does not accept afterpay.",
}
`;

exports[`getUnit cases: RESULT 5`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "order": Object {
      "unitId": "UNIT ID",
    },
    "unit": Object {
      "id": "UNIT ID 3",
      "isAcceptingOrders": true,
      "orderPaymentPolicy": "afterpay",
    },
  },
}
`;

exports[`getUnit cases: RESULT 6`] = `
Object {
  "_tag": "Left",
  "left": "Payment mode is not provided in the order, and the unit does not accept afterpay.",
}
`;

exports[`getUnit cases: RESULT 7`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "order": Object {
      "paymentMode": Object {},
      "unitId": "UNIT ID",
    },
    "unit": Object {
      "id": "UNIT ID 5",
      "isAcceptingOrders": true,
    },
  },
}
`;

exports[`handleRkeeperOrderCases cases: RESULT 1`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "order": Object {
      "unitId": "THE ORDER UNIT ID",
    },
    "unit": Object {
      "id": "NOT AN RKEEPER UNIT",
    },
  },
}
`;

exports[`handleRkeeperOrderCases cases: RESULT 2`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "order": Object {
      "unitId": "THE ORDER UNIT ID",
    },
    "unit": Object {
      "id": "AN RKEEPER UNIT",
      "pos": Object {
        "type": "rkeeper",
      },
    },
  },
}
`;

exports[`handleRkeeperOrderCases cases: RESULT 3`] = `
Object {
  "_tag": "Left",
  "left": "RKEEPER ERROR",
}
`;

exports[`handleRkeeperOrderCases cases: RKEEPER DEPS CALLS 1`] = `Array []`;

exports[`handleRkeeperOrderCases cases: RKEEPER DEPS CALLS 2`] = `
Array [
  Array [
    Object {
      "axiosInstance": Object {},
      "currentTimeISOString": [Function],
      "uuidGenerator": [Function],
    },
  ],
]
`;

exports[`handleRkeeperOrderCases cases: RKEEPER DEPS CALLS 3`] = `
Array [
  Array [
    Object {
      "axiosInstance": Object {},
      "currentTimeISOString": [Function],
      "uuidGenerator": [Function],
    },
  ],
]
`;

exports[`handleRkeeperOrderCases cases: RKEEPER FUNC CALLS 1`] = `Array []`;

exports[`handleRkeeperOrderCases cases: RKEEPER FUNC CALLS 2`] = `
Array [
  Array [
    Object {
      "id": "AN RKEEPER UNIT",
      "pos": Object {
        "type": "rkeeper",
      },
    },
    Object {
      "unitId": "THE ORDER UNIT ID",
    },
  ],
]
`;

exports[`handleRkeeperOrderCases cases: RKEEPER FUNC CALLS 3`] = `
Array [
  Array [
    Object {
      "id": "AN RKEEPER UNIT",
      "pos": Object {
        "type": "rkeeper",
      },
    },
    Object {
      "unitId": "THE ORDER UNIT ID",
    },
  ],
]
`;

exports[`placeOrderCases cases: Case: "All good", PUT CALLS 1`] = `
Array [
  Array [
    Object {
      "Item": Object {
        "archived": false,
        "createdAt": "2022-02-01T00:00:00.000Z",
        "currentStatus": "none",
        "id": "THE UUID",
        "statusLog": Array [
          Object {
            "status": "none",
            "ts": 1627909024677,
            "userId": undefined,
          },
        ],
        "unitId": "THE ORDER UNIT ID",
        "updatedAt": "2022-02-01T00:00:00.000Z",
      },
      "TableName": "ORDER TABLE NAME",
    },
  ],
]
`;

exports[`placeOrderCases cases: Case: "All good", RESULT 1`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "order": Object {
      "archived": false,
      "createdAt": "2022-02-01T00:00:00.000Z",
      "currentStatus": "none",
      "id": "THE UUID",
      "statusLog": Array [
        Object {
          "status": "none",
          "ts": 1627909024677,
          "userId": undefined,
        },
      ],
      "unitId": "THE ORDER UNIT ID",
      "updatedAt": "2022-02-01T00:00:00.000Z",
    },
  },
}
`;

exports[`placeOrderCases cases: Case: "Error happened", PUT CALLS 1`] = `
Array [
  Array [
    Object {
      "Item": Object {
        "archived": false,
        "createdAt": "2022-02-01T00:00:00.000Z",
        "currentStatus": "none",
        "id": "THE UUID",
        "statusLog": Array [
          Object {
            "status": "none",
            "ts": 1627909024677,
            "userId": undefined,
          },
        ],
        "unitId": "THE ORDER UNIT ID",
        "updatedAt": "2022-02-01T00:00:00.000Z",
      },
      "TableName": "ORDER TABLE NAME",
    },
  ],
]
`;

exports[`placeOrderCases cases: Case: "Error happened", RESULT 1`] = `
Object {
  "_tag": "Left",
  "left": "DB ERRORED",
}
`;

exports[`placeOrderCases cases: Case: "Order ID specified", PUT CALLS 1`] = `
Array [
  Array [
    Object {
      "Item": Object {
        "archived": false,
        "createdAt": "2022-02-01T00:00:00.000Z",
        "currentStatus": "none",
        "id": "USER GIVEN ORDER ID",
        "statusLog": Array [
          Object {
            "status": "none",
            "ts": 1627909024677,
            "userId": undefined,
          },
        ],
        "unitId": "THE ORDER UNIT ID",
        "updatedAt": "2022-02-01T00:00:00.000Z",
      },
      "TableName": "ORDER TABLE NAME",
    },
  ],
]
`;

exports[`placeOrderCases cases: Case: "Order ID specified", RESULT 1`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "order": Object {
      "archived": false,
      "createdAt": "2022-02-01T00:00:00.000Z",
      "currentStatus": "none",
      "id": "USER GIVEN ORDER ID",
      "statusLog": Array [
        Object {
          "status": "none",
          "ts": 1627909024677,
          "userId": undefined,
        },
      ],
      "unitId": "THE ORDER UNIT ID",
      "updatedAt": "2022-02-01T00:00:00.000Z",
    },
  },
}
`;
